synapse/lib/kv.py

```diff
@@ -0,0 +1,466 @@
+import os
+import lmdb
+import functools
+
+import synapse.eventbus as s_eventbus
+
+import synapse.lib.msgpack as s_msgpack
+
+
+class KvDict:
+    '''
+    A KvDict uses the KvStor to implement a pythonic dict-like object.
+
+    Unlike the KvLook object, the KvDict keeps all items in the dictionary
+    in memory, so retrieval is fast; and only updates needs to be written
+    to the the underlying KvStor object.
+    '''
+    def __init__(self, stor, iden):
+        self.stor = stor
+        self.iden = iden
+        self.vals = {}
+
+        # load all the props
+        for lkey, lval in stor.iterKvProps(iden):
+            prop = lkey[16:].decode('utf8')
+            self.vals[prop] = s_msgpack.un(lval)
+
+    def items(self):
+        '''
+        Yield (prop, valu) tuples from the KvDict.
+
+        Notes:
+            This yields data from the internal dictionary ``items()`` method.
+            Changes made to the KvDict contents during the consumption of the
+            results of the ``KvDict.items()`` generator are not present in the
+            output of this generator.
+
+        Yields:
+            ((str, object)): Tuple of prop, valu pairs.
+        '''
+        for item in tuple(self.vals.items()):
+            yield item
+
+    def set(self, prop, valu):
+        '''
+        Set a property in the KvDict.
+
+        Args:
+            prop (str): The property name.
+            valu (obj): A msgpack compatible value.
+
+        Returns:
+            None
+        '''
+        if self.vals.get(prop) == valu:
+            return
+
+        self.vals[prop] = valu
```

Comment:
@vertexmc  - I spoke about this with @invisig0th a bit this afternoon.

One problem we introduce with locking is that we can end up taking a big performance hit with locking.  LMDB also provides some mutual exclusion around writer transactions too which help with managing those guys.

One solution we discussed was to do thread-local scopes which could keep a transaction open as needed and promote itself to a writer transaction when needed.  However, there is alot of work to do in order to get that right, and we don't really have a use case within Synapse at the moment which would require us to be super-protective of these structures. In many cases, each dude would be a single thread working on his own namespace.

We decided that we should track this as a known issue and address it down the road, since it is not something we need to tackle right now.

Reply:
Yep makes sense. As long as we document that these are not safe to share between threads, its completely fine. I'd rather make the common case run faster.